test/htbl/dscheck_htbl.ml

```diff
@@ -0,0 +1,263 @@
+module Atomic = Dscheck.TracedAtomic
+
+module Dscheck_htbl (Htbl : Htbl_intf.HTBL) = struct
+  open Htbl
+
+  let try_add htbl k = try_add htbl k k
+  let create ?min_buckets () = create ?min_buckets ~hashed_type:(module Int) ()
+
+  let _two_mem () =
+    Atomic.trace ~record_traces:true (fun () ->
+        Random.init 0;
+
+        let htbl = create () in
+        let added = List.init 10 (fun i -> try_add htbl i) in
+        let found1, found2, found3, found4 =
+          (ref false, ref false, ref false, ref false)
+        in
+
+        Atomic.spawn (fun () ->
+            found1 := mem htbl 1;
+            found2 := mem htbl 20);
+        Atomic.spawn (fun () ->
+            found3 := mem htbl 1;
+            found4 := mem htbl 2);
+
+        Atomic.final (fun () ->
+            Atomic.check (fun () ->
+                List.for_all (fun x -> x) added
+                && !found1 && (not !found2) && !found3 && !found4)));
+
+    Dscheck.Trace_tracker.print_traces stdout
+
+  let _two_add () =
+    Atomic.trace (fun () ->
+        Random.init 0;
+        let htbl = create ~min_buckets:8 () in
+        try_add htbl 1 |> ignore;
+        let added1, added2, added3, added4 =
+          (ref false, ref false, ref false, ref false)
+        in
+
+        Atomic.spawn (fun () ->
+            added1 := try_add htbl 1;
+            added2 := try_add htbl 21);
+        Atomic.spawn (fun () ->
+            added4 := try_add htbl 22;
+            added3 := try_add htbl 1);
+
+        Atomic.final (fun () ->
+            Atomic.check (fun () ->
+                (not !added1) && !added2 && (not !added3) && !added4
+                && mem htbl 1 && mem htbl 21 && mem htbl 22)))
+
+  let _two_add_resize () =
+    (* Trying to trigger resize *)
+    let random_offset = Random.int 1000 in
+    for i = 0 to 5 do
```

Comment:
Hmm... Why do we have the outer loop here? (And in some of the other tests.)

Reply:
> The point is to try multiple seeds to make it more probable to encounter a seed that actually triggers a resize as the choice of resizing is still partially made randomly.

This test is so expensive that I think we should just use a single seed that does trigger a resize and run the test once.  We could add a function to the hash table that returns the current number of buckets and use that to verify that a resize happened.

>  BTW, I was wondering if this (randomly deciding if resize is triggered) is still needed as the size computation is cheaper than before.

Actually it is more expensive than in the Picos version.  Both the Picos version and this Saturn version use a scalable counting mechanism.  They both allocate an atomic counter (or two) per domain.   What makes it expensive to read is that it requires reading the atomic counters of other domains.  This is expensive, because those counters are likely not in the cache of the current domain and so reading the counters potentially generates a lot of cache coherency traffic.  It also requires reading one (or two) counter(s) per domain.  The more domains you have, the more expensive it is to read the counter value.